### ISSUE TITLE
feat: make use of semantic pull requests to ensure common commit standard across the repository

### ISSUE THREAD

**Issue by vinayakkulkarni** [mapbox/mapbox-gl-js#9826]
## Motivation

Re: https://github.com/mapbox/mapbox-gl-js/pull/9811
First of all, thanks for mapbox-gl-js, we've been using mapbox heavily in our organisation.
The motivation behind having commit lint/semantic-pr & following conventional commit messages is just to ensure that it's easy for everyone (bots included) to a common standard.

> `commitlint` helps your team adhering to a commit convention. By supporting npm-installed configurations it makes sharing of commit conventions easy.

## Design Alternatives

You'd need to install the probot app for [semantic pull request](https://probot.github.io/apps/semantic-pull-requests/) and update the branch protection rules by ensuring the commit messages and pull request titles adhere to the conventions.

## Design

It basically follows the conventional config (which is also used by Angular)

### Mock-Up

NA

### Concepts

Basically, if all contributors adhere to the convention and we use semantic versioning, the changelog would automatically be generated based on the `type` of PR, eg. feat: <some feature> would add a new feature entry in the Changelog, making life easy for the releaser.

### Implementation

Refer the PR #9811

**Comment by andrewharvey** [comment]
> Basically, if all contributors adhere to the convention and we use semantic versioning, the changelog would automatically be generated based on the type of PR, eg. feat: would add a new feature entry in the Changelog, making life easy for the releaser.

This is how it works currently through the label attached to the PR, the PR checklist https://github.com/mapbox/mapbox-gl-js/blob/main/.github/PULL_REQUEST_TEMPLATE.md has:

>  apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'

which is used to automatically build the changelog.

My 2c as a casual contributor is labels are much easier to work with than trying to make commit/pr messages machine readable in GitHub.

**Comment by arindam1993** [comment]
I'm all in for this, while labels are easy to work with, they are not enforced and our changelog drafting tool is very flaky so we basically write everything manually which takes a lot of time.
Though this needs more buy-in from the rest of @mapbox/gl-js , actually rejected this idea same time last year.

**Comment by andrewharvey** [comment]
How would this work, since we use squash and rebase to merge commits, it's usually the PR title which is assigned the commit message. If a new contributor makes a bunch of commits which all end up being squashed at merge and isn't sure how to correctly classify it, it's much easier to just change the PR title than try to amend their root commit.

If this only checks the PR title / squashed and rebased commit title I think that'll provide a smoother contributor experience.

Also in the referenced PR

> allowMergeCommits: true

Currently this project does squash and rebase, which personally I've found much cleaner and easier to work with than merge commits.

**Comment by vinayakkulkarni** [comment]
I highly recommend using something like [`semantic-release`](https://github.com/semantic-release/semantic-release) or for even more control, algolia's [`ship.js`](https://community.algolia.com/shipjs/guide/) (personal fav) for auto generating changelog & doing auto-releases instead of the headache of manually writing changelogs etc.

For all above mentioned release-automation tools to work smoothly, you require the commits to follow a specific standard so that they are picked up by those apps and auto-generated changelogs (& releases) are done successfully.

**Comment by arindam1993** [comment]
I don't think we're doing fully-CI generated changelogs anytime soon, we often edit the language of the changelog manually a lot since we also post these changelogs to our blogs and we want them to be a bit more story-like.

We'd want to use this locally to generate a "close-enough" first draft though ( kind of like running `shipjs prepare`)
